debug/debug.go

```diff
@@ -0,0 +1,36 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debug // import "go.opentelemetry.io/otel/debug"
+
+import (
+	"log"
+	"os"
+
+	"github.com/go-logr/logr"
+	"github.com/go-logr/stdr"
+
+	"go.opentelemetry.io/otel/internal/debug"
+)
+
+// SetLogger configures the logger used internally to opentelemetry.
+func SetLogger(logger logr.Logger) {
```

Comment:
What do you think of changing the `logger` type to an interface which would be "compatible" with `logr`?

Pros:
- we could remove the coupling to `logr`. The user could inject its own logger created using `logr` or use its own logger with API similar to `logr`.
- we always have the possibility to create our own implementation with better performance than `logr`

Cons:
- can lead to worse performance b/c of `interface{}` usage (would need to be measured if it is really a problem) [more](https://github.com/go-logr/logr#why-is-this-not-a-pure-interface)

### Background

Here is a great article by @nblumhardt where he describes similar problems and approaches for .NET: https://nblumhardt.com/2016/04/which-logging-abstraction-should-i-use/

Reply:
While I would like to, I can't actually make an interface that is 100% decoupled and uses all of the features of the logr logger.

For example I can't have the [WithValues()](https://pkg.go.dev/github.com/go-logr/logr#Logger.WithValues).  Because they return a `logr.Logger`, and we couldn't we would just return our internal Logger (`debug.Logger`).  

This would then limit this interface to only the `Error()` and `Info()` methods.

We could of course build our own internal logger struct, and have an interface for the logr to our struct, but that would make it that much more awkward to use something like `zap` or `loggrus` because our users would have to wrap those loggers in logr and then in our struct.

Reply:
> While I would like to, I can't actually make an interface that is 100% decoupled and uses all of the features of the logr logger. For example I can't have the WithValues(). Because they return a logr.Logger, and we couldn't we would just return our internal Logger (debug.Logger).

True. But do we want all want the whole API and all of the features of `logr`? I would rather prefer to have a minimal API surface. E.g. hardcode the logging levels and just create an example adapter for the `logr` (e.g. in contrib repo) so that we get other adapters out-of-the-box with minimal effort.

> We could of course build our own internal logger struct, and have an interface for the logr to our struct, but that would make it that much more awkward to use something like zap or loggrus because our users would have to wrap those loggers in logr and then in our struct.

Yeah, true... That is what I meant when I was writing "For users using it, adding an interface is just an unnecessary layer of abstraction.". Still, I think it might be worth comparing and experimenting with.

_I will try to create a PR but I just need to find some time to do it..._

Reply:
Going to try and cover a few topics here:

**Size of the `logr` API**
Right now the `logr` API is about twice as big as I see us using, with `Error`, `Info`, `V`, `WithValues`, and maybe `WithName` being the biggest candidate.  That being said we only over expose this API internally, so we can change it internally if we need to. The customer facing API consists of just a SetLogger.  While we can't update this if we did want to move to an internal implementation we could write an adapter to the new format, and expose a new API that allows directly setting the new logger.

**Internal implementation**
The biggest downside I see to us implementing our own internal implementation is setting up logging goes from:
```
zapLog, _:= zap.NewDevelopment()
debug.SetLogger(zapr.NewLogger(zapLog))
```
to
```
zapLog, _:= zap.NewDevelopment()
zaprLog := zapr.NewLogger(zapLog)
debug.SetLogger(otelLog.NewLogger(zaprLog))
```
This is additional complexity that doesn't seem to gain us much.  The current plan would just to have a transition layer from our internal semantics of levels to the appropiate V(lvl) within logr.

**Conventions**
I was going to try and punt on this as much as possible, but lets hash a bit of this out.  You asked in a different thread about enforcing conventions, eg calling V(100).  First off I think we need to try to instrument our code more before we set hard conventions for new code and think about enforcement of them.  I can already see a need for three different logging levels e.g. 

- Info, things that would print rarely to inform user of the general configuration/state of the system config would be a good candidate
- verbose, things that would readily inform the user of ongoing operations of system, like a report when a batcher sent a batch
- debug, extra details of EVERYTHING that is happening within the system.

The main problem with this is that even within the `*r` world there isn't a convention of what and how `V(lvl)` is represented within the actual loggers (`zap`, `loggrus`, `zerolog`).  What this means is that how to enable different verbosity levels between loggers, and what the convention is in that logger is different.  My suggestion to solve this is to build examples with two or three different loggers to give users an easy onboarding experience.

Reply:
An update on this thread:

** Size of the `logr` API **
I unexported the interal Log, and reexposed it interally via `Info`, `Error`, and `Debug`.  This can allow us to control usage through out the project.  It will not decouple us for `logr`.  We save a lot of complexity by not creating our own interface, and mainting bridges to other logging libraries.

** Conventions **
I addressed it in the main thread, but the TLDR is some exploratory work gave me confidance that there is some uniformity in the logging space, and we can collapse down to three levels.

Reply:
I like it as it decreases the coupling and increases maintainablilty.